### ISSUE TITLE
Investigate macOS tabs

### ISSUE THREAD

**Issue by jwilm** [alacritty/alacritty#1544]
Per @pcwalton's [comment], this may actually be somewhat easy to support.

[comment]: https://news.ycombinator.com/item?id=18007940

**Comment by paulrouget** [comment]
I have some experience in this area. We ran into the same issues with Servo.

On a general note, for tabs, menu, popups (etc) support or any proper integration with MacOS, we tried:

1. create an xcode + obj-c/swift project that would load Servo as a library (no winit/glutin)
2. create a xib the we would in Rust (see https://github.com/paulrouget/servoshell, no winit/glutin)
3. use the [native handle exposed by winit](https://docs.rs/winit/0.11.3/winit/os/index.html) and dynamically build the menu and all the missing MacOS integration bits

*2* was my favorite approach as it didn't require writing Obj-c/Swift code, but we could still design the UI with XCode. But we lose winit events which was great for the multiplatform abstraction. It's probably possible now to tell winit to re-use an existing window (the xib loader creates its own window iirc).

*3* works. It's not optimal as everything needs to be done manually.

I think the best long term approach is to add these features to winit.

As for tabs, specifically, both for Mac and Windows (window support is still in beta, see Redstone 5 release), it's just matter of supporting multiple windows and flagging them a "tab-able", and the OS will do the rest.

Hope that helps.

**Comment by jwilm** [comment]
Thanks for sharing your experience here @paulrouget! Even better that there's a big example program to learn from. :smile:

I hope you don't mind; I have a few questions.

1. When you say "adding these features to winit", are you calling out xib loader/initialization for a winit window?
2. In servoshell, did you reimplement event handling on top of the handle returned by the xib loader?
3. Reflecting on this a bit further, it sounds like servo isn't using winit any longer. Is that true? What major features are you missing from winit today beyond xib support?

cc @francesca64 who might be interested in this discussion

**Comment by paulrouget** [comment]
> 1. When you say "adding these features to winit", are you calling out xib loader/initialization for a winit window?

No. I never used xib + winit.

I meant adding methods to the winit builder to support tabs and menus.

> 2. In servoshell, did you reimplement event handling on top of the handle returned by the xib loader?

Yes. The xib loader will instantiate whatever classes is referenced in the Xib. These classes are implemented in rust. See for example the NSView which holds the gl buffer: https://github.com/paulrouget/servoshell/blob/master/src/platform/cocoa/view.rs - you'll see the `keyUp:`, `scrollWheel:` methods.

> 3. Reflecting on this a bit further, it sounds like servo isn't using winit any longer. Is that true? What major features are you missing from winit today beyond xib support?

We use winit on desktop (but not on mobile).

The ServoShell project has been abandoned. It was just a prototype to experiment with xcode/xib. I haven't worked on it for a year now.

If you plan to just add tab and menu support, implementing these 2 features in winit is probably the way to go. If you want a more advanced UI, implementing a way to make Winit load a xib for you would be pretty amazing.

**Comment by sodiumjoe** [comment]
this sounds like it depends on https://github.com/jwilm/alacritty/issues/607

**Comment by Spirarel** [comment]
Is anyone able to weigh in where tabs are at? The conversation ended up split over several issues.
Last I saw, [macOS window grouping required all windows to belong to the same process.](https://github.com/jwilm/alacritty/issues/607#issuecomment-423746720)

Are native macOS tabs still on the table for Alacritty or has this been abandoned?

**Comment by nixpulvis** [comment]
As I understand the goals of Alacritty, this wouldn't be ruled out if it only requires a few custom macOS additions to support. However if it conflicts with our own ideas of how we want to support something like this across platforms then it would be best to just implement tabs the one way from the start. The main thing I'd like to avoid personally is two implementations of tabs.

**Comment by nicolashahn** [comment]
@nixpulvis Is that on the roadmap? I'm loving alacritty on OSX except for having to use tmux for tabs and splits.

**Comment by herrbischoff** [comment]
Same here. My only gripe is missing tabs. I love the way tabs are implemented in https://github.com/kovidgoyal/kitty, which is DOS-style text tabs at the bottom:

<img width="681" alt="Screenshot 2019-05-07 at 23 32 14" src="https://user-images.githubusercontent.com/8957346/57334582-71951a80-7120-11e9-9a48-5cd36ef4ba8e.png">

I believe this would fit the simplicity requirement quite nicely and would translate cross-platform.

**Comment by chrisduerr** [comment]
This feature is only about the native tab support by the macOS window manager, not about tabs inside of Alacritty.

Those are out of scope for Alacritty and better handled by tmux or the WM/Compositor.

**Comment by herrbischoff** [comment]
I see. Just interpreted @nixpulvis comment

> However if it conflicts with our own ideas of how we want to support something like this across platforms then it would be best to just implement tabs the one way from the start.

as possibly pointing in the direction of cross-platform tabs. If tabs are still generally out of scope, wouldn't it be more consistent to allow multiple terminal windows/instances in macOS instead? Just wondering.

**Comment by chrisduerr** [comment]
> wouldn't it be more consistent to allow multiple terminal windows/instances in macOS instead

That's the goal. Based on the information I have, it's necessary on macOS to spawn separate windows from the same process to have them connected together properly. Once that's the case, it will also allow easily making use of the native tab support.

These issues are all linked together as far as I know, however I'm unaware of any specifics (which is why this issue is labeled as *investigate* macos tabs. If it should turn out that Alacritty will need to do any kind of rendering adjustments to display tabs itself, it will not be supported. However it seems like macOS takes care of that, so it shouldn't be a massive change after proper instancing for macOS is resolved.

**Comment by gilescope** [comment]
This would be so helpful - Alacritty is excellent, but throw in tmux and vi and copy and paste on OSX becomes an amazing symphony of competing configurations. There are 15 blogs devoted to how one might solve this but not using tmux would make things much simpler - at the moment figuring out copy+paste is like watching inception...

**Comment by aste4** [comment]
I'd like to chime in with my reason why tabs are necessary to have in alacritty. 

I use tmux daily. It's my life. The one use case alacritty doesn't solve very well is when I want to use tmux locally but also remotely. With almost any other terminal out there, I can create a new instance of a terminal in one tab, run tmux locally there and create window/panes locally. In a second tab, I can ssh to the remote server and use tmux there. I can switch between tab 1 and tab 2 with some key binding. 

The other similar use case is that if I use tmux locally, ssh into a remote machine and launch vim, certain key presses (like Ctrl-J and Ctrl-K) are not received by the remote vim session because they are captured by the local instance of tmux and instead move up/down to the next pane. I solve this again by creating a new terminal, sshing, and attaching or creating a new tmux session there. 

Supporting these use cases is not impossible but it's very impractical. If sending prefix command to a remote tmux session while in tmux is doable but is a big cognitive tax

One possibility is to have multiple instances of alacritty but from what I've seen in issues raised here, its not so easy/obvious/clean. 

Like @herrbischoff mentioned, if alacritty could have the same text-based tabs that kitty has, alacritty would be a no brainer for me. Or just another tmux like pane that adds an extra layer/dimension. I'd drop my current terminals in a heartbeat. I've successfully run alacritty on both mac and linux and I love how easy it is to configure. 

This terminal is my number 1 choice and I will use it when my day to day doesn't involve working both remotely and locally but right now the lack of tabs is killing me. 

I've tried kitty but the font rendering seems completely off to what I would expect. Alacritty's is perfect with the exception of a font smoothing issue fixed by a patch found in one of the issues here. 

I have zero experience with rust but I am slowly starting to learn in hopes of at least adding this functionality in a fork. I have absolutely zero idea where to begin but if there's no hope for tabs in alacritty then I'd better get right to it. Any pointers on where to look to adding a text based like tab at the top similar to what kitty has?

**Comment by chrisduerr** [comment]
@aste4 Tabs just aren't something that should be supported in a terminal emulator. It's just a waste of effort since it's very complex and completely unrelated to the issues a terminal emulator actually needs to solve. Tiling and tabbing are prime examples of exactly what window managers are there for and if you feel the desire to tab your windows, that's probably where this problem should be raised.

In Linux that's already widely available with many window managers, so it's just about picking one that supports the features you want.

On macOS it's possible to let the macOS WM handle it if the original comment by pcwalton is accurate. That's also something we can do, spawning multiple connected Alacritty processes is something we can do, managing windows is something the macOS WM can do for us then.

On Windows things are the most dire. They actually planned on adding this and I'd assume it would have worked great with Alacritty. I'm not sure if this was just too drastic of a UI/UX change for Windows or what caused this, but they've scrapped it for now. So chances are Windows users are SoL (at least for now), but that's unfortunately a common issue with Windows and I don't think it's Alacritty's shop to solve this.

**Comment by BarryMode** [comment]
I've caught myself several times already hitting ⌘+T trying to open additional tabs. It's something I have done frequently in Terminal where I'll put in a command and know that it's going to take a while to finish, so I'll open a new tab and run multiple processes at the same time. That's where I see the usefulness in it. If there's a way to do this without tabs, I'd love to know but as far as I know, you have to do one command at a time without tabs or the ability to spawn multiple windows.

**Comment by chrisduerr** [comment]
@BarryMode Your shell should support job control, which allows you to put certain commands in the background and keep stdin available. Though most users that rely on this frequently will likely prefer tmux or screen.

**Comment by nixpulvis** [comment]
The discussion here should not be focused on whether or not tabs are useful, I think we all know they are. The question is what application's job it is to implement it. As we've now seen in #2786, tabs can be implemented from a third party tool named `tabbed` on X11 (I've tested it, and it works quite well). If we can do the same kind of thing for macOS without maintaining either a huge macOS specific library, or effectively our own `tabbed` implementation.

**Comment by gabeio** [comment]
Can I just request if tabbing is added that it is optional like config setting to get rid of them or even better yet not bother loading that (module?) into memory. I currently love alacritty as it stands today. I actually appreciate that it helped me get into `tmux` because of the lack of tabs.

**Comment by nixpulvis** [comment]
@ganeio, with the ways things work in X11 on Linux, you get this for free, and only launch a "tabbed" Alacritty session with `tabbed alacritty --embed`. However, if to support macOS there is considerable dynamic linking required, it may be less desirable. I hope there's a solution as clean as what we seem to have with X11.

**Comment by pcwalton** [comment]
Tab support is built into AppKit, which Alacritty is already linking to. I don't know much of the details of how it works, however. There will likely need to be some macOS-specific code, though it could be encapsulated in a `macos-tabs` crate or something like that.

**Comment by gabeio** [comment]
I'm totally cool with even needing to rebuild alacritty (I already build alacritty currently) to disable the tabs so long as it's an easy flag option to make, build environment variable or something similar 😎

**Comment by chrisduerr** [comment]
Yeah, this issue is purely about investigating the macOS tabs @pcwalton is talking about. It's not about Windows or X11, those will be completely unaffected. And there should not be any rendering code added to Alacritty.

So there's no need to worry about having to explicitly disable Alacritty's tabs, since Alacritty itself will not have tabs. We're just integrating functionality provided by the window manager.

This is also why it's a bit unfortunate that Windows seems to have canceled their tabs feature. Because that makes it the only platform where this isn't possible. But with Windows' terminal support there are many issues that Windows users currently just have to live with because of lacking support from Microsoft.

**Comment by pcwalton** [comment]
The main difficulty would involve enabling multi-window support for Alacritty. Allowing for macOS tabs would be trivial after that.

**Comment by chrisduerr** [comment]
Yeah, that describes the TL;DR on this issue pretty well. And that's also what has been stated so far. Currently this means that this issue is blocked by #607.

**Comment by harrysolovay** [comment]
Any movement on this issue? Lack of multiple tabs/windows support it the only feature keeping me from using Alacritty.

**Comment by gilescope** [comment]
Windows Terminal has tabs and works pretty well. I will use that while waiting for a rust terminal that can do tabs.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by adzierzanowski** [comment]
I'm using XQuartz on macOS and I compiled `tabbed`. It works fine with `xterm`. Even though `alacritty` opens inside X server if XQuartz is running, `tabbed alacritty --embed` creates detached windows instead of tabs. I wonder what is the point of failure here.

**Comment by kchibisov** [comment]
@adzierzanowski oh it won't work, since X11 is limited to linux/bsd(mean not macOS).

**Comment by diego-lipinski-de-castro** [comment]
I found alacritty today and will already need to stop using because it does not have tabs support, oh well

**Comment by adzierzanowski** [comment]
@diego-lipinski-de-castro Try `tmux`, it supports tabs, keyboard shortcuts and has mouse support.

<img width="722" alt="Screen Shot 2020-09-18 at 18 21 03" src="https://user-images.githubusercontent.com/7986275/93621395-c4046d80-f9db-11ea-861b-be8f8eb53f31.png">

**Comment by coolaj86** [comment]
> Tabs just aren't something that should be supported in a terminal emulator.

That's a moral argument. Not that we can't have morals, but... why can't we have tabs?

I mean, I just can't imagine a person saying "Tabs just aren't something that should be supported in a web browser" and yet, the Terminal is the browser of the computer.

**Comment by coolaj86** [comment]
Could there be some sort of agreement on a way to put tabs in, but keep that in a separate fork that can easily track without other conflicts?

**Comment by coolaj86** [comment]
@adzierzanowski How does one get <kbd>Cmd ⌘</kbd> + <kbd>T</kbd> and Click to switch tab working in `tmux`?

**Comment by coolaj86** [comment]
@adzierzanowski How does one get <kbd>Cmd ⌘</kbd> + <kbd>T</kbd> and Click to switch tab working in `tmux`?

**Comment by jbg** [comment]
@coolaj86 please be considerate of the fact that your issue comments here are emailed to probably >50 people. You can get a `tmux` tutorial elsewhere.

**Comment by coolaj86** [comment]
@jbg I believe that people come to this issue through search, looking for a solution. If there's a solution and someone here knows it, let them speak.

Likewise, let the people who don't want to follow this thread unsubscribe.

Much like it's not a terminal's job to manage tabs, it's not my job to micro-manage who should be exposed to what kind of information.

**Comment by SalomonSmeke** [comment]
Excuse me for adding to the noise.

+1 to tmux, but it's a really popular tool and you're likely going to find many comprehensive or quick guides with a quick google.

This issue is blocked by another issue, about multiple window support. It's listed above in the buried comments!

I also think this would be a great feature; but it's blocked. Even if it would be given priority, that would have to happen first. Might be worth marking that other issue with comments mentioning the issue remains relevant; but don't spam it. Please. I'm sure the maintainers have things they do.

Maybe try a hand at it yourself! I'm down to donate some time if you find folk familiar with this codebase that need a monkey.

**Comment by Jeppelelle** [comment]
Any news on this? Apple have made it fairly easy to setup tabs in applications that are running NSWindow

I am not to familiar with Rust (creating NSWindow that allow for tabs is super easy in Xcode + Swift, basically just a bool for allowing/implementing tabs & some minor extra code for controller logic if needed, usually not needed since macOS/X11/Quartz/whatever takes care if it for you), having said that, it should not really be that much of an issue for anyone familiar to Rust to implement tabs on macOS:

Alacritty already uses flags for target_os = "macos" for creating windows with WindowBuilder, going by the Rust docs WindowBuilder / Window in Rust implements platformspecific (not with winit since it seems somewhat limited to which flags can be set on NSWindows, would probably have to go through WindowExt/WindowBuilderExt, or any other way of accessing NSWindows directly): https://docs.rs/winit/0.15.1/i686-apple-darwin/winit/os/macos/trait.WindowExt.html

So, there should be an easy access to NSWindows, as many have stated in this thread, some have mentioned NSWindowTab which works but **allowsAutomaticWindowTabbing** is even easier to setup, it takes a bool (true) which is basically a "set it and forget it" toggle; the system takes care of most of the rest (most, because as i said, sometimes developers want custom logic to tabs & windows but i think this works good enough for most macOS Alacritty users who are longing for tabs) https://developer.apple.com/documentation/appkit/nswindow/1646657-allowsautomaticwindowtabbing

So, Alacritty DON'T need to implement its own tab system on macOS, it just have to allow NSWindow to implement native macOS tabs which the system takes care of outside of it's apps (when intercommunication between tabs is not needed, which Alacritty don't need, the default emulator (zfs) don't) & multiwindows would not be affected with this implementation, user/system decides if it wants new multiple windows or opening a new tab inside a window (basically with automatic tabbing macOS just groups together multiple windows inside a tabbed NSWindow, all Alacritty instances would still be separate)

So yeah, a Tab inside ActionWrapper that implements a bool toggle from "allowsAutomaticWindowTabbing" on macOS would be awesome

**Comment by onedr0p** [comment]
I don't see this mentioned anywhere but [Wezterm](https://github.com/wez/wezterm) has tab support and is also written in Rust. Maybe the developers can see how it's done there?

**Comment by kchibisov** [comment]
>I don't see this mentioned anywhere in this issue but Wezterm has tab support (not using OS APIs afaik) and is also written in Rust. Maybe the developers can see how it's done there?

It doesn't matter how it's done there, and it has nothing to do with rust. The only thing that matters is someone dedicated to work on that feature on macOS, since core maintainers are not using macOS. However the one thing that we need to do before doing macOS tabs is multi window support.

**Comment by chrisduerr** [comment]
> Any news on this?

If anyone else is curious about this question, please just read the previous discussions. There is no news about this, because even the issue blocking it isn't done yet. So unless that changes there's no point at all in asking again and again.

Chances are this will take a long time too, because there's no benefit to any of the main contributors to Alacritty. Nobody contributing to Alacritty actually wants to use macOS. If you're familiar with Rust and willing to look into this, I'm always happy to mentor people through the implementation process, but there's no incentive for most contributors to work on this and constantly asking about it doesn't really help.

**Comment by terhechte** [comment]
Just wanted to chime in. Although this is in theory rather easy to do on macOS and even with `winit` and `glutin`, it is much harder for Alacritty because `allowsAutomaticWindowTabbing` works on a per window basis. Each window (`NSWindow`) becomes a new tab. Alacritty, however, doesn't support multiple windows. When you open another *window* it actually forks a new process / instance. macOS / Cocoa doesn't support tabbing for multiple processes. See `event::spawn_new_instance`. 

I'd guess rewriting Alacritty to support multiple event loops (one per window) in one process is a lot more work. Actually, I briefly looked at the code and couldn't see an easy way to pull this off 😔

**Comment by chichid** [comment]
> I don't see this mentioned anywhere in this issue but [Wezterm](https://github.com/wez/wezterm) has tab support (not using OS APIs afaik) and is also written in Rust. Maybe the developers can see how it's done there?

Thanks for the hint, Wezterm is definitely the best option to get the best of both worlds: a fast GPU accelerated terminal and a lot better feature support.

**Comment by xenoterracide** [comment]
I like alacritty, especially the vi like navigation, I was hoping for tabs that might be like vim's tabs and once I'm in that vi mode I could just do `:gt` `:gT` `:tabnew`, not really wanting native support. I'll have to take a new look at tmux, but in some ways I'm wondering if I've got too much vi inception going on...

**Comment by xenoterracide** [comment]
I already miss Alacritty's vi controls, honestly the kind of tabs I'd want from alacritty are like those of vim's so I can use the same controls to switch tabs.

**Comment by Sleepful** [comment]
> Tabs just aren't something that should be supported in a terminal emulator. It's just a waste of effort since it's very complex and completely unrelated to the issues a terminal emulator actually needs to solve

Uh, I mean, it's fine if you don't want to do it, but *why not?*
Tmux acts as a wall between the terminal and the apps, keyboard input and graphical output is limited to the scan-codes that tmux likes.

**Comment by sstubbs** [comment]
Firstly thank you for this amazing project. I've been using https://github.com/zellij-org/zellij with alacritty for tabs and it's working quite well so far.

**Comment by chrisduerr** [comment]
macOS tab support now works on macOS. There's currently a bug in winit which might require manual resize sometimes but that is tracked separately.

**Comment by kchibisov** [comment]
@cmoog you can use tabs now if you go to system preferences and use prefer tabs for new window option. Each new window created with CreateWindow will tab.

**Comment by nixpulvis** [comment]
> Each new window created with CreateWindow will tab.

Just to be precise, the `action` is `CreateNewWindow`.

On the macOS side, it still seems somewhat unfortunate that I can't choose between new windows and new tabs. I've never heard of this whole "Perfer tabs" nonsense but I guess I'll set it to "always" since I can then drag the tab out into a window, whereas I don't think I can drag a window back into a tab.

If anyone knows more about this, I'd be curious to hear more.

**Comment by chrisduerr** [comment]
> If anyone knows more about this, I'd be curious to hear more.

Yeah, if there's an option in the macOS API to set this preference on a per-application level, that might be something we can provide to make things a little easier.

**Comment by nixpulvis** [comment]
Honestly, I just want to try and understand the motivation for this setting in macOS in the first place. Why the hell would I want a new window to "prefer" anything? Opening a new window should be (pretty much) deterministic, no? Sure there's always going to be conflicts if multiple processes are trying to interact with the WM at once (or maybe someone has clever solutions to this too?), but I see nothing wrong with having a <kbd>⌘ + N</kbd> and a separate <kbd>⌘ + T</kbd> binding for these two different actions.

I'm guessing this is yet another example of iOS crap leaking into macOS. Mobile gestures are much more limited than what we are used to with a proper keyboard. Frankly, I'd be happy relearning my mobile experience around a keyboard long before I give up my desktop/laptop keyboard "power-user" experience with reliable and various keybindings.

But yea, if Alacritty can just force macOS to always open in tabs regardless of preference, that would be nice. At least that way people will be *able* to use tabs and windows without having to learn about this insane configuration.

**Comment by chrisduerr** [comment]
> Honestly, I just want to try and understand the motivation for this setting in macOS in the first place. Why the hell would I want a new window to "prefer" anything? Opening a new window should be (pretty much) deterministic, no? Sure there's always going to be conflicts if multiple processes are trying to interact with the WM at once (or maybe someone has clever solutions to this too?), but I see nothing wrong with having a ⌘ + N and a separate ⌘ + T binding for these two different actions.

I mean it is deterministic I think? If you prefer tabs, it'll always open as tabs?

> But yea, if Alacritty can just force macOS to always open in tabs regardless of preference, that would be nice. At least that way people will be able to use tabs and windows without having to learn about this insane configuration.

Should likely be a config option, there have been multiple people that do **not** want this (myself included).

**Comment by nixpulvis** [comment]
If it's configurable from within Alacritty then perhaps we could have both a `CreateNewWindow` and `CreateNewTab` action... sadly, this is super leaky, since other platforms aren't going to work the same way. So, this falls into the realm of winit et. all.

Anyways, I completely agree with you. Always opening in tabs is not right either.

**Comment by Jeppelelle** [comment]
> So my `CreateNewWindow` binding still creates a new window rather than a tab despite having "Prefer tabs" set to always in Settings. Any idea why that might be? I'm running macOS 12.2.1

It seems to only work if you have window decoration set to full, check that setting

**Comment by Noctem** [comment]
> It seems to only work if you have window decoration set to full, check that setting
Ahh, that was it. Thank you.

**Comment by chetan-reddy** [comment]
FYI the keyboard shortcuts to switch between tabs are Cmd+Shift+[ and Cmd+Shift+]
I could not find a keyboard shortcut to switch to a tab that's not adjacent to the current tab.
Thanks for implementing this.

**Comment by scott2b** [comment]
> > So my `CreateNewWindow` binding still creates a new window rather than a tab despite having "Prefer tabs" set to always in Settings. Any idea why that might be? I'm running macOS 12.2.1
> 
> It seems to only work if you have window decoration set to full, check that setting

What does this mean?

**Comment by ardnew** [comment]
> @cmoog you can use tabs now if you go to system preferences and use prefer tabs for new window option. Each new window created with CreateWindow will tab.

Thank you so much for solving this! It's a ~pane~pain moving terminal windows around with all decorations disabled!

Dunno if you or I have the latest macOS (probably you), but my option is named slightly differently in `System Preferences > General`:
![no-tabs](https://user-images.githubusercontent.com/3837367/167490518-963f4f13-7cc6-486f-912a-19817866a2f3.jpg)

**Comment by Samasaur1** [comment]
@chrisduerr is there any objection to Alacritty supporting a macOS-only action `CreateNewTab` if winit were to support it? To clarify, I don't mean having Alacritty implement its own tabs, I mean using the macOS system window manager.

(This is of course an issue with the operating system, but if winit were to support it, it doesn't feel unreasonable for Alacritty to support it as well, and from what I can tell seems like it would be pretty simple from the Alacritty side)

**Comment by chrisduerr** [comment]
I'm not aware of macOS' API with regards to spawning things as tabs. But I wouldn't dismiss it immediately.

**Comment by kchibisov** [comment]
The upstream macOS PR is here https://github.com/rust-windowing/winit/pull/2552

**Comment by Samasaur1** [comment]
That's actually not what I was talking about, because I didn't see that PR until I put some work into modifying `winit` and `alacritty` to support tabs. I'm working on cleaning up my changes to submit as PRs (I'll put the diffs in a collapsible section, though, in case anyone wants to patch their own), but I now have a working version of Alacritty with new tabs and some of the "normal" macOS keyboard shortcuts for interacting with them:

https://user-images.githubusercontent.com/30577766/204153752-113e681d-f646-42e1-a2f3-af20216c73b3.mp4

https://user-images.githubusercontent.com/30577766/204153753-a8dbf5d3-c297-4526-bb03-6438592e2fdd.mp4

<details>
  <summary><code>alacritty</code> diff (from <code>d92a8a0e16eda3df7566e5c995aa0d258b6b76f7</code>)</summary>

```diff
diff --git a/Cargo.lock b/Cargo.lock
index 25471d1a..37c5424e 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2,6 +2,22 @@
 # It is not intended for manual editing.
 version = 3
 
+[[package]]
+name = "ab_glyph"
+version = "0.2.18"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4dcdbc68024b653943864d436fe8a24b028095bc1cf91a8926f8241e4aaffe59"
+dependencies = [
+ "ab_glyph_rasterizer",
+ "owned_ttf_parser",
+]
+
+[[package]]
+name = "ab_glyph_rasterizer"
+version = "0.1.7"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "330223a1aecc308757b9926e9391c9b47f8ef2dbd8aea9df88312aea18c5e8d6"
+
 [[package]]
 name = "adler"
 version = "1.0.2"
@@ -39,7 +55,7 @@ dependencies = [
  "serde_yaml",
  "unicode-width",
  "wayland-client",
- "windows-sys",
+ "windows-sys 0.36.1",
  "winit",
  "x11-dl",
  "xdg",
@@ -92,9 +108,33 @@ dependencies = [
  "signal-hook-mio",
  "unicode-width",
  "vte",
- "windows-sys",
+ "windows-sys 0.36.1",
 ]
 
+[[package]]
+name = "android-activity"
+version = "0.4.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4165a1aef703232031b40a6e8908c2f9e314d495f11aa7f98db75d39a497cc6a"
+dependencies = [
+ "android-properties",
+ "bitflags",
+ "cc",
+ "jni-sys",
+ "libc",
+ "log",
+ "ndk",
+ "ndk-context",
+ "ndk-sys",
+ "num_enum",
+]
+
+[[package]]
+name = "android-properties"
+version = "0.2.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "fc7eb209b1518d6bb87b283c20095f5228ecda460da70b44f0802523dea6da04"
+
 [[package]]
 name = "arrayref"
 version = "0.3.6"
@@ -142,6 +182,25 @@ version = "0.1.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "0d8c1fef690941d3e7788d328517591fecc684c084084702d6ff1641e993699a"
 
+[[package]]
+name = "block-sys"
+version = "0.1.0-beta.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "0fa55741ee90902547802152aaf3f8e5248aab7e21468089560d4c8840561146"
+dependencies = [
+ "objc-sys",
+]
+
+[[package]]
+name = "block2"
+version = "0.2.0-alpha.6"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "8dd9e63c1744f755c2f60332b88de39d341e5e86239014ad839bd71c106dec42"
+dependencies = [
+ "block-sys",
+ "objc2-encode",
+]
+
 [[package]]
 name = "bumpalo"
 version = "3.11.0"
@@ -172,6 +231,9 @@ name = "cc"
 version = "1.0.73"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "2fff2a6927b3bb87f9595d67196a70493f627687a71d87a0d692242c33f58c11"
+dependencies = [
+ "jobserver",
+]
 
 [[package]]
 name = "cfg-if"
@@ -422,41 +484,6 @@ version = "0.2.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "b365fabc795046672053e29c954733ec3b05e4be654ab130fe8f1f94d7051f35"
 
-[[package]]
-name = "darling"
-version = "0.13.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a01d95850c592940db9b8194bc39f4bc0e89dee5c4265e4b1807c34a9aba453c"
-dependencies = [
- "darling_core",
- "darling_macro",
-]
-
-[[package]]
-name = "darling_core"
-version = "0.13.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "859d65a907b6852c9361e3185c862aae7fafd2887876799fa55f5f99dc40d610"
-dependencies = [
- "fnv",
- "ident_case",
- "proc-macro2",
- "quote",
- "strsim",
- "syn",
-]
-
-[[package]]
-name = "darling_macro"
-version = "0.13.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9c972679f83bdf9c42bd905396b6c3588a843a17f0f16dfcfa3e2c5d57441835"
-dependencies = [
- "darling_core",
- "quote",
- "syn",
-]
-
 [[package]]
 name = "dirs"
 version = "4.0.0"
@@ -544,7 +571,7 @@ dependencies = [
  "cfg-if 1.0.0",
  "libc",
  "redox_syscall",
- "windows-sys",
+ "windows-sys 0.36.1",
 ]
 
 [[package]]
@@ -693,7 +720,7 @@ dependencies = [
  "once_cell",
  "raw-window-handle 0.5.0",
  "wayland-sys 0.30.0-beta.12",
- "windows-sys",
+ "windows-sys 0.36.1",
  "x11-dl",
 ]
 
@@ -704,7 +731,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "2c3c95a2d7a54bab0c74759794efb5cd63470d4504cbe85ed4114dc82c98bdc1"
 dependencies = [
  "gl_generator",
- "windows-sys",
+ "windows-sys 0.36.1",
 ]
 
 [[package]]
@@ -747,12 +774,6 @@ dependencies = [
  "libc",
 ]
 
-[[package]]
-name = "ident_case"
-version = "1.0.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b9e0384b61958566e926dc50660321d12159025e767c18e043daf26b70104c39"
-
 [[package]]
 name = "indexmap"
 version = "1.9.1"
@@ -816,6 +837,15 @@ version = "0.3.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "8eaf4bc02d17cbdd7ff4c7438cafcdf7fb9a4613313ad11b4f8fefe7d3fa0130"
 
+[[package]]
+name = "jobserver"
+version = "0.1.25"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "068b1ee6743e4d11fb9c6a1e6064b3693a1b600e7f5f5988047d98b3dc9fb90b"
+dependencies = [
+ "libc",
+]
+
 [[package]]
 name = "js-sys"
 version = "0.3.60"
@@ -997,7 +1027,7 @@ dependencies = [
  "libc",
  "log",
  "wasi",
- "windows-sys",
+ "windows-sys 0.36.1",
 ]
 
 [[package]]
@@ -1077,35 +1107,6 @@ version = "0.1.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "27b02d87554356db9e9a873add8782d4ea6e3e58ea071a9adb9a2e8ddb884a8b"
 
-[[package]]
-name = "ndk-glue"
-version = "0.7.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0434fabdd2c15e0aab768ca31d5b7b333717f03cf02037d5a0a3ff3c278ed67f"
-dependencies = [
- "libc",
- "log",
- "ndk",
- "ndk-context",
- "ndk-macro",
- "ndk-sys",
- "once_cell",
- "parking_lot 0.12.1",
-]
-
-[[package]]
-name = "ndk-macro"
-version = "0.3.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0df7ac00c4672f9d5aece54ee3347520b7e20f158656c7db2e6de01902eb7a6c"
-dependencies = [
- "darling",
- "proc-macro-crate",
- "proc-macro2",
- "quote",
- "syn",
-]
-
 [[package]]
 name = "ndk-sys"
 version = "0.4.0"
@@ -1216,6 +1217,32 @@ dependencies = [
  "objc_id",
 ]
 
+[[package]]
+name = "objc-sys"
+version = "0.2.0-beta.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "df3b9834c1e95694a05a828b59f55fa2afec6288359cda67146126b3f90a55d7"
+
+[[package]]
+name = "objc2"
+version = "0.3.0-beta.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "fe31e5425d3d0b89a15982c024392815da40689aceb34bad364d58732bcfd649"
+dependencies = [
+ "block2",
+ "objc-sys",
+ "objc2-encode",
+]
+
+[[package]]
+name = "objc2-encode"
+version = "2.0.0-pre.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "abfcac41015b00a120608fdaa6938c44cb983fee294351cc4bac7638b4e50512"
+dependencies = [
+ "objc-sys",
+]
+
 [[package]]
 name = "objc_id"
 version = "0.1.1"
@@ -1237,6 +1264,15 @@ version = "6.3.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "9ff7415e9ae3fff1225851df9e0d9e4e5479f947619774677a63572e55e80eff"
 
+[[package]]
+name = "owned_ttf_parser"
+version = "0.17.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "18904d3c65493a9f0d7542293d1a7f69bfdc309a6b9ef4f46dc3e58b0577edc5"
+dependencies = [
+ "ttf-parser",
+]
+
 [[package]]
 name = "parking_lot"
 version = "0.11.2"
@@ -1282,7 +1318,7 @@ dependencies = [
  "libc",
  "redox_syscall",
  "smallvec",
- "windows-sys",
+ "windows-sys 0.36.1",
 ]
 
 [[package]]
@@ -1471,11 +1507,11 @@ checksum = "d29ab0c6d3fc0ee92fe66e2d99f700eab17a8d57d1c1d3b748380fb20baa78cd"
 
 [[package]]
 name = "sctk-adwaita"
-version = "0.4.3"
+version = "0.5.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "61270629cc6b4d77ec1907db1033d5c2e1a404c412743621981a871dc9c12339"
+checksum = "6be6d8911b40c5f91332ce27d2458d57fffc0b061816bf847df23a16f9d3615f"
 dependencies = [
- "crossfont",
+ "ab_glyph",
  "log",
  "smithay-client-toolkit",
  "tiny-skia",
@@ -1727,6 +1763,12 @@ dependencies = [
  "serde",
 ]
 
+[[package]]
+name = "ttf-parser"
+version = "0.17.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "375812fa44dab6df41c195cd2f7fecb488f6c09fbaafb62807488cefab642bff"
+
 [[package]]
 name = "unicode-ident"
 version = "1.0.4"
@@ -2012,51 +2054,106 @@ version = "0.36.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "ea04155a16a59f9eab786fe12a4a450e75cdb175f9e0d80da1e17db09f55b8d2"
 dependencies = [
- "windows_aarch64_msvc",
- "windows_i686_gnu",
- "windows_i686_msvc",
- "windows_x86_64_gnu",
- "windows_x86_64_msvc",
+ "windows_aarch64_msvc 0.36.1",
+ "windows_i686_gnu 0.36.1",
+ "windows_i686_msvc 0.36.1",
+ "windows_x86_64_gnu 0.36.1",
+ "windows_x86_64_msvc 0.36.1",
+]
+
+[[package]]
+name = "windows-sys"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "5a3e1820f08b8513f676f7ab6c1f99ff312fb97b553d30ff4dd86f9f15728aa7"
+dependencies = [
+ "windows_aarch64_gnullvm",
+ "windows_aarch64_msvc 0.42.0",
+ "windows_i686_gnu 0.42.0",
+ "windows_i686_msvc 0.42.0",
+ "windows_x86_64_gnu 0.42.0",
+ "windows_x86_64_gnullvm",
+ "windows_x86_64_msvc 0.42.0",
 ]
 
+[[package]]
+name = "windows_aarch64_gnullvm"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "41d2aa71f6f0cbe00ae5167d90ef3cfe66527d6f613ca78ac8024c3ccab9a19e"
+
 [[package]]
 name = "windows_aarch64_msvc"
 version = "0.36.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "9bb8c3fd39ade2d67e9874ac4f3db21f0d710bee00fe7cab16949ec184eeaa47"
 
+[[package]]
+name = "windows_aarch64_msvc"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "dd0f252f5a35cac83d6311b2e795981f5ee6e67eb1f9a7f64eb4500fbc4dcdb4"
+
 [[package]]
 name = "windows_i686_gnu"
 version = "0.36.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "180e6ccf01daf4c426b846dfc66db1fc518f074baa793aa7d9b9aaeffad6a3b6"
 
+[[package]]
+name = "windows_i686_gnu"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "fbeae19f6716841636c28d695375df17562ca208b2b7d0dc47635a50ae6c5de7"
+
 [[package]]
 name = "windows_i686_msvc"
 version = "0.36.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "e2e7917148b2812d1eeafaeb22a97e4813dfa60a3f8f78ebe204bcc88f12f024"
 
+[[package]]
+name = "windows_i686_msvc"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "84c12f65daa39dd2babe6e442988fc329d6243fdce47d7d2d155b8d874862246"
+
 [[package]]
 name = "windows_x86_64_gnu"
 version = "0.36.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "4dcd171b8776c41b97521e5da127a2d86ad280114807d0b2ab1e462bc764d9e1"
 
+[[package]]
+name = "windows_x86_64_gnu"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "bf7b1b21b5362cbc318f686150e5bcea75ecedc74dd157d874d754a2ca44b0ed"
+
+[[package]]
+name = "windows_x86_64_gnullvm"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "09d525d2ba30eeb3297665bd434a54297e4170c7f1a44cad4ef58095b4cd2028"
+
 [[package]]
 name = "windows_x86_64_msvc"
 version = "0.36.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "c811ca4a8c853ef420abd8592ba53ddbbac90410fab6903b3e79972a631f7680"
 
+[[package]]
+name = "windows_x86_64_msvc"
+version = "0.42.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "f40009d85759725a34da6d89a94e63d7bdc50a862acf0dbc7c8e488f1edcb6f5"
+
 [[package]]
 name = "winit"
 version = "0.27.5"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bb796d6fbd86b2fd896c9471e6f04d39d750076ebe5680a3958f00f5ab97657c"
 dependencies = [
+ "android-activity",
  "bitflags",
- "cocoa",
  "core-foundation",
  "core-graphics",
  "dispatch",
@@ -2065,10 +2162,8 @@ dependencies = [
  "log",
  "mio 0.8.4",
  "ndk",
- "ndk-glue",
- "objc",
+ "objc2",
  "once_cell",
- "parking_lot 0.12.1",
  "percent-encoding",
  "raw-window-handle 0.4.3",
  "raw-window-handle 0.5.0",
@@ -2079,7 +2174,7 @@ dependencies = [
  "wayland-client",
  "wayland-protocols",
  "web-sys",
- "windows-sys",
+ "windows-sys 0.42.0",
  "x11-dl",
 ]
 
diff --git a/alacritty.yml b/alacritty.yml
index 4e9de637..d904aabf 100644
--- a/alacritty.yml
+++ b/alacritty.yml
@@ -110,9 +110,9 @@
   #multiplier: 3
 
 # Font configuration
-#font:
+font:
   # Normal (roman) font face
-  #normal:
+  normal:
     # Font family
     #
     # Default:
@@ -120,6 +120,8 @@
     #   - (Linux/BSD) monospace
     #   - (Windows) Consolas
     #family: monospace
+    family: JetBrainsMonoNL Nerd Font Mono
+    style: Regular
 
     # The `style` can be specified to pick a specific face.
     #style: Regular
@@ -131,6 +133,8 @@
     # If the bold family is not specified, it will fall back to the
     # value specified for the normal font.
     #family: monospace
+    # family: JetBrainsMonoNL Nerd Font Mono
+    # style: Bold
 
     # The `style` can be specified to pick a specific face.
     #style: Bold
@@ -142,6 +146,8 @@
     # If the italic family is not specified, it will fall back to the
     # value specified for the normal font.
     #family: monospace
+    # family: JetBrainsMonoNL Nerd Font Mono
+    # style: Italic
 
     # The `style` can be specified to pick a specific face.
     #style: Italic
@@ -153,12 +159,15 @@
     # If the bold italic family is not specified, it will fall back to the
     # value specified for the normal font.
     #family: monospace
+    # family: JetBrainsMonoNL Nerd Font Mono
+    # style: Bold Italic
 
     # The `style` can be specified to pick a specific face.
     #style: Bold Italic
 
   # Point size
   #size: 11.0
+  size: 12.0
 
   # Offset is the extra space around each character. `offset.y` can be thought
   # of as modifying the line spacing, and `offset.x` as modifying the letter
@@ -758,7 +767,7 @@
 #
 # If the same trigger is assigned to multiple actions, all of them are executed
 # in the order they were defined in.
-#key_bindings:
+key_bindings:
   #- { key: Paste,                                       action: Paste          }
   #- { key: Copy,                                        action: Copy           }
   #- { key: L,         mods: Control,                    action: ClearLogNotice }
@@ -869,6 +878,11 @@
   #- { key: F,              mods: Command|Control,            action: ToggleFullscreen      }
   #- { key: F,              mods: Command, mode: ~Search,     action: SearchForward         }
   #- { key: B,              mods: Command, mode: ~Search,     action: SearchBackward        }
+  - { key: N, mods: Command, action: CreateNewWindow }
+  - { key: N, mods: Command|Shift, action: SpawnNewInstance }
+  - { key: T, mods: Command, action: CreateNewTab }
+  - { key: Tab, mods: Control, action: SelectNextTab }
+  - { key: Tab, mods: Control|Shift, action: SelectPreviousTab }
 
 #debug:
   # Display the time it takes to redraw each frame.
diff --git a/alacritty/Cargo.toml b/alacritty/Cargo.toml
index 307b8c8c..25b4031a 100644
--- a/alacritty/Cargo.toml
+++ b/alacritty/Cargo.toml
@@ -30,7 +30,8 @@ serde = { version = "1", features = ["derive"] }
 serde_yaml = "0.8"
 serde_json = "1"
 glutin = { version = "0.30.0", default-features = false, features = ["egl", "wgl"] }
-winit = { version = "0.27.4", default-features = false, features = ["serde"] }
+# winit = { version = "0.27.4", default-features = false, features = ["serde"] }
+winit = { path = "/Users/sam/Desktop/tabbed-alacritty/winit", default-features = false, features = ["serde"] }
 notify-debouncer-mini = { version = "0.2.1", default-features = false }
 parking_lot = "0.12.0"
 crossfont = { version = "0.5.0", features = ["force_system_fontconfig"] }
diff --git a/alacritty/src/config/bindings.rs b/alacritty/src/config/bindings.rs
index 7b654b66..5b5c38a3 100644
--- a/alacritty/src/config/bindings.rs
+++ b/alacritty/src/config/bindings.rs
@@ -183,6 +183,15 @@ pub enum Action {
     /// Create a new Alacritty window.
     CreateNewWindow,
 
+    #[cfg(target_os = "macos")]
+    CreateNewTab,
+
+    #[cfg(target_os = "macos")]
+    SelectNextTab,
+
+    #[cfg(target_os = "macos")]
+    SelectPreviousTab,
+
     /// Toggle fullscreen.
     ToggleFullscreen,
 
diff --git a/alacritty/src/display/window.rs b/alacritty/src/display/window.rs
index 93e83677..49d5a66a 100644
--- a/alacritty/src/display/window.rs
+++ b/alacritty/src/display/window.rs
@@ -202,6 +202,22 @@ impl Window {
         })
     }
 
+    #[cfg(target_os = "macos")]
+    pub fn add_window_as_tab(&self, tab: &Window) {
+        self.window.add_window_as_tab(&tab.window);
+        // self.window.add_window_as_tab(tab.window.ns_window());
+    }
+
+    #[cfg(target_os = "macos")]
+    pub fn select_next_tab(&self) {
+        self.window.select_next_tab();
+    }
+
+    #[cfg(target_os = "macos")]
+    pub fn select_previous_tab(&self) {
+        self.window.select_previous_tab();
+    }
+
     #[inline]
     pub fn raw_window_handle(&self) -> RawWindowHandle {
         self.window.raw_window_handle()
diff --git a/alacritty/src/event.rs b/alacritty/src/event.rs
index 7cf77096..6ae74dbc 100644
--- a/alacritty/src/event.rs
+++ b/alacritty/src/event.rs
@@ -96,6 +96,12 @@ pub enum EventType {
     Message(Message),
     Scroll(Scroll),
     CreateWindow(WindowOptions),
+    #[cfg(target_os = "macos")]
+    CreateTab(WindowOptions),
+    #[cfg(target_os = "macos")]
+    SelectNextTab,
+    #[cfg(target_os = "macos")]
+    SelectPreviousTab,
     #[cfg(unix)]
     IpcConfig(IpcConfig),
     BlinkCursor,
@@ -415,6 +421,26 @@ impl<'a, N: Notify + 'a, T: EventListener> input::ActionContext<T> for ActionCon
             .send_event(Event::new(EventType::CreateWindow(WindowOptions::default()), None));
     }
 
+    #[cfg(target_os = "macos")]
+    fn create_new_tab(&mut self) {
+        let mut options = WindowOptions::default();
+        if let Ok(working_directory) = foreground_process_path(self.master_fd, self.shell_pid) {
+            options.terminal_options.working_directory = Some(working_directory);
+        }
+
+        let _ = self.event_proxy.send_event(Event::new(EventType::CreateTab(options), None));
+    }
+
+    #[cfg(target_os = "macos")]
+    fn select_next_tab(&mut self) {
+        let _  = self.event_proxy.send_event(Event::new(EventType::SelectNextTab, None));
+    }
+
+    #[cfg(target_os = "macos")]
+    fn select_previous_tab(&mut self) {
+        let _  = self.event_proxy.send_event(Event::new(EventType::SelectPreviousTab, None));
+    }
+
     fn spawn_daemon<I, S>(&self, program: &str, args: I)
     where
         I: IntoIterator<Item = S> + Debug + Copy,
@@ -1170,6 +1196,12 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                 #[cfg(unix)]
                 EventType::IpcConfig(_) => (),
                 EventType::ConfigReload(_) | EventType::CreateWindow(_) => (),
+                #[cfg(target_os = "macos")]
+                EventType::CreateTab(_) => (),
+                #[cfg(target_os = "macos")]
+                EventType::SelectNextTab => (),
+                #[cfg(target_os = "macos")]
+                EventType::SelectPreviousTab => (),
             },
             WinitEvent::RedrawRequested(_) => *self.ctx.dirty = true,
             WinitEvent::WindowEvent { event, .. } => {
@@ -1274,7 +1306,9 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                     | WindowEvent::ThemeChanged(_)
                     | WindowEvent::HoveredFile(_)
                     | WindowEvent::Touch(_)
-                    | WindowEvent::Moved(_) => (),
+                    | WindowEvent::Moved(_)
+                    | WindowEvent::TouchpadMagnify { .. }
+                    | WindowEvent::TouchpadRotate { .. } => (),
                 }
             },
             WinitEvent::Suspended { .. }
@@ -1370,6 +1404,46 @@ impl Processor {
         Ok(())
     }
 
+    #[cfg(target_os = "macos")]
+    pub fn create_tab(
+        &mut self,
+        event_loop: &EventLoopWindowTarget<Event>,
+        proxy: EventLoopProxy<Event>,
+        options: WindowOptions,
+    ) -> Result<(), Box<dyn Error>> {
+        //TODO: get the current window, rather than the whichever window this ends up being
+        //  perhaps iterate through all windows and check if they are active?
+        // let window = self.windows.iter().next().as_ref().unwrap().1;
+        let window = self.windows.iter().find(|(_, context)| context.focused()).unwrap().1;
+        let window_context = window.additional(
+            event_loop,
+            proxy,
+            self.config.clone(),
+            options,
+            #[cfg(all(feature = "wayland", not(any(target_os = "macos", windows))))]
+            self.wayland_event_queue.as_ref(),
+        )?;
+        window.display.window.add_window_as_tab(&window_context.display.window);
+        // window_context.display.window.add_window_as_tab(None);
+
+        self.windows.insert(window_context.id(), window_context);
+        Ok(())
+    }
+
+    #[cfg(target_os = "macos")]
+    pub fn select_next_tab(&mut self) -> Result<(), Box<dyn Error>> {
+        let window = self.windows.iter().find(|(_, context)| context.focused()).unwrap().1;
+        window.display.window.select_next_tab();
+        Ok(())
+    }
+
+    #[cfg(target_os = "macos")]
+    pub fn select_previous_tab(&mut self) -> Result<(), Box<dyn Error>> {
+        let window = self.windows.iter().find(|(_, context)| context.focused()).unwrap().1;
+        window.display.window.select_previous_tab();
+        Ok(())
+    }
+
     /// Run the event loop.
     ///
     /// The result is exit code generate from the loop.
@@ -1523,6 +1597,34 @@ impl Processor {
                         error!("Could not open window: {:?}", err);
                     }
                 },
+                #[cfg(target_os = "macos")]
+                WinitEvent::UserEvent(Event {
+                    payload: EventType::CreateTab(options), ..
+                }) => {
+                    for window_context in self.windows.values_mut() {
+                        window_context.display.make_not_current();
+                    }
+
+                    if let Err(err) = self.create_tab(event_loop, proxy.clone(), options) {
+                        error!("Could not open tab: {:?}", err);
+                    }
+                },
+                #[cfg(target_os = "macos")]
+                WinitEvent::UserEvent(Event {
+                    payload: EventType::SelectNextTab, ..
+                }) => {
+                    if let Err(err) = self.select_next_tab() {
+                        error!("could not select next tab: {:?}", err);
+                    }
+                },
+                #[cfg(target_os = "macos")]
+                WinitEvent::UserEvent(Event {
+                    payload: EventType::SelectPreviousTab, ..
+                }) => {
+                    if let Err(err) = self.select_previous_tab() {
+                        error!("could not select previous tab: {:?}", err);
+                    }
+                },
                 // Process events affecting all windows.
                 WinitEvent::UserEvent(event @ Event { window_id: None, .. }) => {
                     for window_context in self.windows.values_mut() {
diff --git a/alacritty/src/input.rs b/alacritty/src/input.rs
index c8d948ff..43bf2026 100644
--- a/alacritty/src/input.rs
+++ b/alacritty/src/input.rs
@@ -82,6 +82,12 @@ pub trait ActionContext<T: EventListener> {
     fn terminal_mut(&mut self) -> &mut Term<T>;
     fn spawn_new_instance(&mut self) {}
     fn create_new_window(&mut self) {}
+    #[cfg(target_os = "macos")]
+    fn create_new_tab(&mut self) {}
+    #[cfg(target_os = "macos")]
+    fn select_next_tab(&mut self) {}
+    #[cfg(target_os = "macos")]
+    fn select_previous_tab(&mut self) {}
     fn change_font_size(&mut self, _delta: f32) {}
     fn reset_font_size(&mut self) {}
     fn pop_message(&mut self) {}
@@ -344,6 +350,12 @@ impl<T: EventListener> Execute<T> for Action {
             Action::ClearLogNotice => ctx.pop_message(),
             Action::SpawnNewInstance => ctx.spawn_new_instance(),
             Action::CreateNewWindow => ctx.create_new_window(),
+            #[cfg(target_os = "macos")]
+            Action::CreateNewTab => ctx.create_new_tab(),
+            #[cfg(target_os = "macos")]
+            Action::SelectNextTab => ctx.select_next_tab(),
+            #[cfg(target_os = "macos")]
+            Action::SelectPreviousTab => ctx.select_previous_tab(),
             Action::ReceiveChar | Action::None => (),
         }
     }
diff --git a/alacritty/src/window_context.rs b/alacritty/src/window_context.rs
index 833586ea..768efde6 100644
--- a/alacritty/src/window_context.rs
+++ b/alacritty/src/window_context.rs
@@ -505,6 +505,10 @@ impl WindowContext {
         self.display.window.id()
     }
 
+    pub fn focused(&self) -> bool {
+        self.terminal.lock().is_focused
+    }
+
     /// Write the ref test results to the disk.
     pub fn write_ref_test_results(&self) {
         // Dump grid state.
```
</details>

<details>
  <summary><code>winit</code> diff (from <code>1786c877ec418372500fcdcac1eeda6a04eefc5b</code>)</summary>

```diff
diff --git a/src/platform_impl/macos/appkit/window.rs b/src/platform_impl/macos/appkit/window.rs
index 7f859746..d27c190f 100644
--- a/src/platform_impl/macos/appkit/window.rs
+++ b/src/platform_impl/macos/appkit/window.rs
@@ -207,6 +207,15 @@ extern_methods!(
 
         #[sel(sendEvent:)]
         pub unsafe fn sendEvent(&self, event: &NSEvent);
+    
+        #[sel(addTabbedWindow:ordered:)]
+        pub unsafe fn addTabbedWindow(&self, window: &NSWindow, ordered: isize);
+
+        #[sel(selectNextTab:)]
+        pub unsafe fn selectNextTab(&self, sender: Option<&Object>);
+
+        #[sel(selectPreviousTab:)]
+        pub unsafe fn selectPreviousTab(&self, sender: Option<&Object>);
     }
 );
 
diff --git a/src/platform_impl/macos/util/async.rs b/src/platform_impl/macos/util/async.rs
index f7135206..ead4cd8e 100644
--- a/src/platform_impl/macos/util/async.rs
+++ b/src/platform_impl/macos/util/async.rs
@@ -93,6 +93,14 @@ pub(crate) fn set_ignore_mouse_events(window: &NSWindow, ignore: bool) {
     });
 }
 
+pub(crate) fn add_window_as_tab(window: &NSWindow, tab: &NSWindow, ordering: isize) {
+    let window = unsafe { MainThreadSafe(mem::transmute::<&NSWindow, &'static NSWindow>(window)) };
+    let tab = unsafe { MainThreadSafe(mem::transmute::<&NSWindow, &'static NSWindow>(tab)) };
+    Queue::main().exec_async(move || {
+        unsafe { window.addTabbedWindow(*tab, ordering); }
+    });
+}
+
 // `toggleFullScreen` is thread-safe, but our additional logic to account for
 // window styles isn't.
 pub(crate) fn toggle_full_screen_async(
diff --git a/src/platform_impl/macos/window.rs b/src/platform_impl/macos/window.rs
index 4a420108..5546f504 100644
--- a/src/platform_impl/macos/window.rs
+++ b/src/platform_impl/macos/window.rs
@@ -773,6 +773,11 @@ impl WinitWindow {
         );
     }
 
+    #[inline]
+    pub(crate) fn add_window_as_tab(&self, window: &WinitWindow, ordered: isize) {
+        util::add_window_as_tab(self, window, ordered);
+    }
+
     #[inline]
     pub(crate) fn fullscreen(&self) -> Option<Fullscreen> {
         let shared_state_lock = self.lock_shared_state("fullscreen");
diff --git a/src/window.rs b/src/window.rs
index 442084ff..73d2a3f1 100644
--- a/src/window.rs
+++ b/src/window.rs
@@ -851,6 +851,19 @@ impl Window {
         self.window.set_window_level(level)
     }
 
+    #[inline]
+    pub fn add_window_as_tab(&self, tab: &Window) {
+        self.window.add_window_as_tab(&tab.window, 1);
+    }
+
+    pub fn select_next_tab(&self) {
+        unsafe { self.window.selectNextTab(None); };
+    }
+
+    pub fn select_previous_tab(&self) {
+        unsafe { self.window.selectPreviousTab(None); };
+    }
+
     /// Sets the window icon.
     ///
     /// On Windows and X11, this is typically the small icon in the top-left
```
</details>

**Comment by ernie** [comment]
@Samasaur1 This looks amazing! Thanks for your work on this, and I sincerely hope it gets merged!

**Comment by KenjiTakahashi** [comment]
Good stuff @Samasaur1! I went ahead and also implemented selecting Nth tab on top of your changes: https://github.com/KenjiTakahashi/alacritty/tree/select_nth_tab.
[In Alacritty, I expose `SelectTab0` to `SelectTab9` actions ATM, as I don't see any existing way for actions to take parameters. This is what the standard Terminal.app is doing by default as well.]

**Comment by Samasaur1** [comment]
The latest commit on [my `async_macos_tabs` branch](https://github.com/Samasaur1/alacritty/tree/async-macos-tabs) supports @KenjiTakahashi's changes, and I also added an action to select the _last_ tab (the default behavior of ⌘9 on macOS). Anyone interested in running this version should be able to build it by running the following:
```
git clone -b async_macos_tabs Samasaur1/alacritty
cd alacritty
make app
```
(you'll need Rust/Cargo installed)

***

Unfortunately, I'm running into some issues merging these changes into `alacritty` and `winit`. In between the version of `winit` used by the latest versions of `alacritty` and the latest version of `winit`, they switched their macOS NSWindow calls to be synchronous, and all the builds of `alacritty` I've made using this new version of winit freeze on tab-related actions (e.g., new tab). So the version I'm using, and the one I provided instructions for building above, still uses the older version of `winit`. And the changes to `alacritty` can't be merged unless the changes to `winit` have already been merged. So at the moment, until I have more time to investigate these hangs, I'm at an impasse. But the build above should work!

**Comment by sfsam** [comment]
FWIW, I also took a stab at adding native tabs. This version doesn't modify `winit`. I just used `objc::msg_send` from within Alacritty. I'm new to Rust and Alacritty and might not know what I'm doing. :)

Tabs can be added with `Cmd-T` and navigation with `Cmd-Shift-[`, `Cmd-Shift-]`, and `Cmd-<1-9>` seem to work. 

In case anyone is interested, the code is here: https://github.com/sfsam/alacritty/commits/dev

**Comment by kchibisov** [comment]
@sfsam there's an upstream PR in winit for stuff like that and I think in general it should go there https://github.com/rust-windowing/winit/pull/2552.

Exposing winit bindings for tabbing is fine with me at least. It could be sort of confusing though, given that none of the other platforms will get that sort of tabbing.

**Comment by nixpulvis** [comment]
This has always confused me a but in the macOS configuration's language. It says 'prefer' tabs, which implies that the application/OS still gets some say.

As a user of macOS/Windows/i3/sway/gnome/kde/etc... how much alacritty configuration would it take to end up with something somewhat consistent across platforms?

**Comment by chrisduerr** [comment]
> As a user of macOS/Windows/i3/sway/gnome/kde/etc... how much alacritty configuration would it take to end up with something somewhat consistent across platforms?

I mean that just entirely depends on your preferred configuration.

Assuming you're talking about tabs mainly, you'd just have to bind everything that comes by default on something like i3 in the application itself. I don't think macOS provides any controls for this.

**Comment by kchibisov** [comment]
Adding a common option you'd expect without configurable bindings on macOS sounds good enough to me. Because you have common settings across the different applications.

**Comment by sfsam** [comment]
> This has always confused me a bit in the macOS configuration's language. It says 'prefer' tabs, which implies that the application/OS still gets some say.

Apps can opt-out of automatic window tabbing even if the user has selected it: 
[-allowsAutomaticWindowTabbing](https://developer.apple.com/documentation/appkit/nswindow/1646657-allowsautomaticwindowtabbing?language=objc) 

They would do this if they don't want the system's one-size-fits-all solution. I think this is pretty common. There are a lot of results here: https://github.com/search?q=allowsAutomaticWindowTabbing&type=code

Mac apps with tabs don't generally rely on the automatic window tabbing feature that is settable in System Preferences for their own tabs. Probably lots of users don't even know this setting. Mac apps implement their own tabbing in a way that makes sense for them. See iTerm2 or MacVim for example. Tabbing behavior isn't an option. It's a standard feature of the apps.

**Comment by kchibisov** [comment]
@sfsam are you interested in upstream some of your patches? Like we have issues for every patch you have in your fork except icon and background clear(I don't understand why you'd need it, I think to color the titlebar?).

I think neovide added that background color clearing thing, but not sure how it'll work with live reload and background changes.

I've added tabbing API upstream in winit, and will add it myself in alacritty, however we have an issue for title bar formatting which you sort of do, but hacky, see https://github.com/alacritty/alacritty/issues/4436 .

**Comment by sfsam** [comment]
@kchibisov You can use anything in my fork that you think is useful.

**Comment by kchibisov** [comment]
@sfsam I was mostly curious behind the motivation for the background color patch, and whether you wanted to upstream the title formatting changes in a way we've outlined. Though, it just knowing the motivation behind the color changes could be enough, I'm not familiar with macOS users workflows that much, so that's why I've asked.

The tabbing macOS feature I wrote based on the patch I've linked before in the winit which is very minimal in alacritty.

**Comment by sfsam** [comment]
@kchibisov The color patch is to make opaque windows render properly on MacOS. The `NSWindow` class needs an opaque fill color (`NSColor.windowBackgroundColor`) to properly draw the window's border and non-translucent title bar. Without the opaque fill, NSWindow doesn't draw a border and the title bar allows objects behind it to 'glow' through. FWIW, the kitty terminal emulator project also just patched this issue: https://github.com/kovidgoyal/kitty/commit/9e0e23ae993ff9c577ee37e51803208a653674db
